File: src/server/kwp/KwpSession.ts
import { EventEmitter } from "node:events";
import { SerialTransport } from "../transport/SerialTransport.js";
import { ElmClient } from "../elm/ElmClient.js";
import { PidReader } from "./PidReader.js";
import { KNOWN_PIDS, TEST_PING_CID } from "./knownPids.js";
import { buildReadByCidRequest, containsPositiveResponseMarker } from "./protocol.js";
import type { ConnectionState, PidReading } from "../../shared/types.js";

const SETTLE_DELAY_MS = 2000;
/** Well under the 2-4s "tester present" ceiling from HANDOVER.md §3.3. */
const POLL_INTERVAL_MS = 1000;

/**
 * Owns the cold-start init sequence (§3.2), the keep-alive/recovery pattern
 * (§3.3), and polling the 5 known PIDs (§3.4). Knows what KWP2000 services
 * and CIDs mean; talks to the ECU only through ElmClient, which has no idea
 * what any of this means in protocol terms.
 */
export class KwpSession extends EventEmitter {
  private state: ConnectionState = "disconnected";
  private transport: SerialTransport | null = null;
  private elm: ElmClient | null = null;
  private pidReader: PidReader | null = null;
  private pollTimer: NodeJS.Timeout | null = null;
  private polling = false;
  private connectedPath: string | null = null;

  getState(): ConnectionState {
    return this.state;
  }

  getPort(): string | null {
    return this.connectedPath;
  }

  async connect(path: string): Promise<void> {
    if (this.state === "connecting" || this.state === "connected") {
      throw new Error(`KwpSession: cannot connect while state is "${this.state}"`);
    }

    this.setState("connecting", `Opening ${path}`);
    const transport = new SerialTransport({ path });
    this.transport = transport;
    this.elm = new ElmClient(transport);
    this.pidReader = new PidReader(this.elm);

    try {
      await transport.open();
      await this.runInitSequence();
      this.connectedPath = path;
      this.setState("connected");
      this.startPolling();
    } catch (err) {
      await transport.close().catch(() => undefined);
      this.transport = null;
      this.elm = null;
      this.pidReader = null;
      this.setState("error", err instanceof Error ? err.message : String(err));
      throw err;
    }
  }

  async disconnect(): Promise<void> {
    this.stopPolling();
    await this.transport?.close().catch(() => undefined);
    this.transport = null;
    this.elm = null;
    this.pidReader = null;
    this.connectedPath = null;
    this.setState("disconnected");
  }

  /** Cold-start init sequence per HANDOVER.md §3.2. */
  private async runInitSequence(): Promise<void> {
    const elm = this.requireElm();

    // The original app sends ATZ twice and doesn't check the response.
    await elm.sendCommand("ATZ", { timeoutMs: 3000 }).catch(() => undefined);
    await elm.sendCommand("ATZ", { timeoutMs: 3000 }).catch(() => undefined);

    await elm.sendCommandUntilOk("ATE0");
    await elm.sendCommandUntilOk("ATH0");
    await elm.sendCommandUntilOk("ATSP5");
    await elm.sendCommandUntilOk("ATSH8101F1");

    await delay(SETTLE_DELAY_MS);

    const ok = await this.sendTestPing();
    if (!ok) {
      throw new Error('KwpSession: init sequence finished but test ping "22111F" got no positive response');
    }
  }

  /** §3.2/§3.3 connectivity check: success = response contains "62". */
  private async sendTestPing(): Promise<boolean> {
    try {
      const raw = await this.requireElm().sendCommand(buildReadByCidRequest(TEST_PING_CID), {
        timeoutMs: 1000,
      });
      return containsPositiveResponseMarker(raw);
    } catch {
      return false;
    }
  }

  /** Keep-alive recovery ladder per §3.3: re-ping -> force fast init -> full restart. */
  private async recover(): Promise<boolean> {
    if (await this.sendTestPing()) return true;

    try {
      await this.requireElm().sendCommand("ATFI", { timeoutMs: 3000 });
    } catch {
      // fall through to the full restart below
    }
    if (await this.sendTestPing()) return true;

    try {
      const elm = this.requireElm();
      await elm.sendCommandUntilOk("ATSP5");
      await elm.sendCommandUntilOk("ATSH8101F1");
    } catch {
      return false;
    }
    return this.sendTestPing();
  }

  private startPolling(): void {
    this.stopPolling();
    this.pollTimer = setInterval(() => {
      void this.pollOnce();
    }, POLL_INTERVAL_MS);
  }

  private stopPolling(): void {
    if (this.pollTimer) {
      clearInterval(this.pollTimer);
      this.pollTimer = null;
    }
  }

  private async pollOnce(): Promise<void> {
    if (this.polling || this.state !== "connected" || !this.pidReader) return;
    this.polling = true;
    try {
      const readings = await this.pidReader.readAll(KNOWN_PIDS);
      this.emit("pids", readings satisfies PidReading[]);

      const sessionDropped = readings.every((reading) => reading.error !== undefined);
      if (sessionDropped) {
        const recovered = await this.recover();
        if (!recovered) {
          this.stopPolling();
          this.setState("error", "Lost KWP session and recovery (§3.3) failed");
        }
      }
    } finally {
      this.polling = false;
    }
  }

  private requireElm(): ElmClient {
    if (!this.elm) throw new Error("KwpSession: not connected");
    return this.elm;
  }

  private setState(state: ConnectionState, detail?: string): void {
    this.state = state;
    this.emit("state", state, detail);
  }
}

function delay(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export declare interface KwpSession {
  on(event: "state", listener: (state: ConnectionState, detail?: string) => void): this;
  on(event: "pids", listener: (readings: PidReading[]) => void): this;
  emit(event: "state", state: ConnectionState, detail?: string): boolean;
  emit(event: "pids", readings: PidReading[]): boolean;
}

File: desktop/ProtonEcuToolkit.Core/Kwp/KwpSession.cs
using ProtonEcuToolkit.Core.Elm;
using ProtonEcuToolkit.Core.Models;
using ProtonEcuToolkit.Core.Transport;

namespace ProtonEcuToolkit.Core.Kwp;

/// <summary>
/// Owns the cold-start init sequence (§3.2), the keep-alive/recovery pattern
/// (§3.3), and polling the 5 known PIDs (§3.4). Knows what KWP2000 services
/// and CIDs mean; talks to the ECU only through ElmClient, which has no idea
/// what any of this means in protocol terms.
/// </summary>
public sealed class KwpSession : IDisposable
{
    private const int SettleDelayMs = 2000;

    /// <summary>Well under the 2-4s "tester present" ceiling from HANDOVER.md §3.3.</summary>
    private const int PollIntervalMs = 1000;

    private ConnectionState _state = ConnectionState.Disconnected;
    private SerialTransport? _transport;
    private ElmClient? _elm;
    private PidReader? _pidReader;
    private CancellationTokenSource? _pollCts;
    private Task? _pollTask;
    private bool _polling;
    private string? _connectedPath;

    public event Action<ConnectionState, string?>? StateChanged;

    public event Action<List<PidReading>>? PidsUpdated;

    public ConnectionState State => _state;

    public string? Port => _connectedPath;

    public async Task ConnectAsync(string path)
    {
        if (_state is ConnectionState.Connecting or ConnectionState.Connected)
        {
            throw new InvalidOperationException($"KwpSession: cannot connect while state is \"{_state}\"");
        }

        SetState(ConnectionState.Connecting, $"Opening {path}");
        var transport = new SerialTransport(path);
        _transport = transport;
        _elm = new ElmClient(transport);
        _pidReader = new PidReader(_elm);

        try
        {
            await transport.OpenAsync().ConfigureAwait(false);
            await RunInitSequenceAsync().ConfigureAwait(false);
            _connectedPath = path;
            SetState(ConnectionState.Connected);
            StartPolling();
        }
        catch (Exception err)
        {
            await SafeCloseAsync(transport).ConfigureAwait(false);
            _transport = null;
            _elm = null;
            _pidReader = null;
            SetState(ConnectionState.Error, err.Message);
            throw;
        }
    }

    public async Task DisconnectAsync()
    {
        await StopPollingAsync().ConfigureAwait(false);
        if (_transport is { } transport)
        {
            await SafeCloseAsync(transport).ConfigureAwait(false);
        }
        _transport = null;
        _elm = null;
        _pidReader = null;
        _connectedPath = null;
        SetState(ConnectionState.Disconnected);
    }

    /// <summary>§3.7 ReadDiagnosticTroubleCodesByStatus. Codes come back as a raw, undecoded hex blob (see Dtc.cs).</summary>
    public Task<DtcActionResult> ScanDtcsAsync() =>
        RunDtcRequestAsync(Dtc.BuildDtcScanRequest(), Dtc.ParseDtcScanResponse);

    /// <summary>§3.7 ClearDiagnosticInformation.</summary>
    public Task<DtcActionResult> ClearDtcsAsync() =>
        RunDtcRequestAsync(Dtc.BuildDtcClearRequest(), Dtc.ParseDtcClearResponse);

    /// <summary>
    /// Sends an arbitrary already-built request hex string and returns the
    /// raw response, for the PID/CID scanner. Has no idea what the bytes
    /// mean - same level of knowledge as ElmClient, just gated on being connected.
    /// </summary>
    public Task<string> SendRawAsync(string requestHex, int timeoutMs) =>
        _state != ConnectionState.Connected
            ? throw new InvalidOperationException($"KwpSession: cannot send while state is \"{_state}\"")
            : RequireElm().SendCommandAsync(requestHex, timeoutMs);

    /// <summary>Stops the regular dashboard poll loop so the scanner has the transport to itself.</summary>
    public Task PausePollingAsync() => _state == ConnectionState.Connected ? StopPollingAsync() : Task.CompletedTask;

    /// <summary>Resumes the regular dashboard poll loop after a scan finishes.</summary>
    public void ResumePolling()
    {
        if (_state == ConnectionState.Connected) StartPolling();
    }

    /// <summary>Cheap re-ping, escalating to the full §3.3 recovery ladder if needed. Used as the scanner's keep-alive check.</summary>
    public Task<bool> EnsureAliveAsync() => RecoverAsync();

    private async Task<DtcActionResult> RunDtcRequestAsync(string requestHex, Func<string, DtcResponse?> parse)
    {
        if (_state != ConnectionState.Connected)
        {
            throw new InvalidOperationException(
                $"KwpSession: cannot send a DTC request while state is \"{_state}\"");
        }

        var timestamp = DateTimeOffset.UtcNow;
        try
        {
            var raw = await RequireElm().SendCommandAsync(requestHex, 2000).ConfigureAwait(false);
            var parsed = parse(raw);
            if (parsed is null)
            {
                return new DtcActionResult(false, raw, null, timestamp, $"Unrecognized response: \"{raw}\"");
            }
            if (!parsed.Positive)
            {
                var nrc = parsed.Nrc is { } nrcValue ? $"0x{nrcValue:x}" : "unknown";
                return new DtcActionResult(false, raw, parsed.Nrc, timestamp, $"Negative response, NRC={nrc}");
            }
            return new DtcActionResult(true, parsed.DataHex, null, timestamp);
        }
        catch (Exception err)
        {
            return new DtcActionResult(false, "", null, timestamp, err.Message);
        }
    }

    /// <summary>Cold-start init sequence per HANDOVER.md §3.2.</summary>
    private async Task RunInitSequenceAsync()
    {
        var elm = RequireElm();

        // The original app sends ATZ twice and doesn't check the response.
        await TryAsync(() => elm.SendCommandAsync("ATZ", 3000)).ConfigureAwait(false);
        await TryAsync(() => elm.SendCommandAsync("ATZ", 3000)).ConfigureAwait(false);

        await elm.SendCommandUntilOkAsync("ATE0").ConfigureAwait(false);
        await elm.SendCommandUntilOkAsync("ATH0").ConfigureAwait(false);
        await elm.SendCommandUntilOkAsync("ATSP5").ConfigureAwait(false);
        await elm.SendCommandUntilOkAsync("ATSH8101F1").ConfigureAwait(false);

        await Task.Delay(SettleDelayMs).ConfigureAwait(false);

        var ok = await SendTestPingAsync().ConfigureAwait(false);
        if (!ok)
        {
            throw new InvalidOperationException(
                "KwpSession: init sequence finished but test ping \"22111F\" got no positive response");
        }
    }

    /// <summary>§3.2/§3.3 connectivity check: success = response contains "62".</summary>
    private async Task<bool> SendTestPingAsync()
    {
        try
        {
            var raw = await RequireElm()
                .SendCommandAsync(Protocol.BuildReadByCidRequest(KnownPids.TestPingCid), 1000)
                .ConfigureAwait(false);
            return Protocol.ContainsPositiveResponseMarker(raw);
        }
        catch
        {
            return false;
        }
    }

    /// <summary>Keep-alive recovery ladder per §3.3: re-ping -&gt; force fast init -&gt; full restart.</summary>
    private async Task<bool> RecoverAsync()
    {
        if (await SendTestPingAsync().ConfigureAwait(false)) return true;

        await TryAsync(() => RequireElm().SendCommandAsync("ATFI", 3000)).ConfigureAwait(false);
        if (await SendTestPingAsync().ConfigureAwait(false)) return true;

        try
        {
            var elm = RequireElm();
            await elm.SendCommandUntilOkAsync("ATSP5").ConfigureAwait(false);
            await elm.SendCommandUntilOkAsync("ATSH8101F1").ConfigureAwait(false);
        }
        catch
        {
            return false;
        }

        return await SendTestPingAsync().ConfigureAwait(false);
    }

    private void StartPolling()
    {
        StopPollingSync();
        var cts = new CancellationTokenSource();
        _pollCts = cts;
        _pollTask = Task.Run(() => PollLoopAsync(cts.Token));
    }

    private async Task PollLoopAsync(CancellationToken token)
    {
        using var timer = new PeriodicTimer(TimeSpan.FromMilliseconds(PollIntervalMs));
        try
        {
            while (await timer.WaitForNextTickAsync(token).ConfigureAwait(false))
            {
                await PollOnceAsync().ConfigureAwait(false);
            }
        }
        catch (OperationCanceledException)
        {
            // expected on disconnect / recovery failure
        }
    }

    private async Task StopPollingAsync()
    {
        if (_pollCts is { } cts)
        {
            cts.Cancel();
            if (_pollTask is { } task)
            {
                await task.ConfigureAwait(false);
            }
            cts.Dispose();
        }
        _pollCts = null;
        _pollTask = null;
    }

    private void StopPollingSync()
    {
        _pollCts?.Cancel();
        _pollCts = null;
        _pollTask = null;
    }

    private async Task PollOnceAsync()
    {
        if (_polling || _state != ConnectionState.Connected || _pidReader is null) return;
        _polling = true;
        try
        {
            var readings = await _pidReader.ReadAllAsync(KnownPids.All).ConfigureAwait(false);
            PidsUpdated?.Invoke(readings);

            var sessionDropped = readings.All(reading => reading.Error is not null);
            if (sessionDropped)
            {
                var recovered = await RecoverAsync().ConfigureAwait(false);
                if (!recovered)
                {
                    StopPollingSync();
                    SetState(ConnectionState.Error, "Lost KWP session and recovery (§3.3) failed");
                }
            }
        }
        finally
        {
            _polling = false;
        }
    }

    private ElmClient RequireElm() => _elm ?? throw new InvalidOperationException("KwpSession: not connected");

    private void SetState(ConnectionState state, string? detail = null)
    {
        _state = state;
        StateChanged?.Invoke(state, detail);
    }

    private static async Task TryAsync(Func<Task> action)
    {
        try
        {
            await action().ConfigureAwait(false);
        }
        catch
        {
            // swallow, mirrors the TS `.catch(() => undefined)` pattern
        }
    }

    private static async Task SafeCloseAsync(SerialTransport transport)
    {
        try
        {
            await transport.CloseAsync().ConfigureAwait(false);
        }
        catch
        {
            // mirrors `.catch(() => undefined)`
        }
    }

    public void Dispose()
    {
        StopPollingSync();
        _transport?.Dispose();
    }
}
